File: backend/app/routes/agent_runs.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query, Request, Response, status
from sqlalchemy.orm import Session

from app.agents.contracts import (
    AgentRunCancelResponse,
    AgentRunCreate,
    AgentRunDeleteResponse,
    AgentRunDetail,
    AgentRunExecuteRequest,
    AgentRunStatus,
    ExecutionMode,
)
from app.api.pagination import (
    SortOrder,
    apply_sort,
    apply_text_search,
    paginate_query,
)
from app.config import settings
from app.database import get_db
from app.models.agent_run import AgentRun
from app.services.agent_execution_service import (
    cancel_agent_run,
    execute_agent_run,
)
from app.services.agent_orchestration_service import (
    create_agent_run,
    delete_agent_run,
    get_agent_run,
    retry_agent_run,
    run_detail,
    serialize_collection_result,
)


router = APIRouter()


@router.post(
    "/",
    response_model=AgentRunDetail,
    status_code=status.HTTP_201_CREATED,
    summary="Plan a persistent multi-agent workflow",
)
def create_run(
    data: AgentRunCreate,
    db: Session = Depends(get_db),
) -> AgentRunDetail:
    return run_detail(create_agent_run(db, data))


@router.get(
    "/",
    summary="List the authenticated user's agent workflows",
)
def list_runs(
    request: Request,
    response: Response,
    db: Session = Depends(get_db),
    search: str | None = Query(
        default=None,
        min_length=1,
        max_length=200,
        description="Search workflow goals, routing reasons, and errors.",
    ),
    run_status: AgentRunStatus | None = Query(default=None, alias="status"),
    execution_mode: ExecutionMode | None = Query(default=None),
    sort_by: str = Query(
        default="created_at",
        pattern="^(id|status|execution_mode|created_at|updated_at)$",
    ),
    sort_order: SortOrder = Query(default=SortOrder.desc),
    page: int | None = Query(default=None, ge=1),
    page_size: int | None = Query(
        default=None,
        ge=1,
        le=settings.api_max_page_size,
    ),
):
    query = db.query(AgentRun)
    query = apply_text_search(
        query,
        search,
        (
            AgentRun.goal,
            AgentRun.routing_reason,
            AgentRun.error_message,
        ),
    )

    if run_status is not None:
        query = query.filter(AgentRun.status == run_status.value)
    if execution_mode is not None:
        query = query.filter(AgentRun.execution_mode == execution_mode.value)

    query = apply_sort(
        query,
        AgentRun,
        sort_by=sort_by,
        sort_order=sort_order,
        allowed_fields=(
            "id",
            "status",
            "execution_mode",
            "created_at",
            "updated_at",
        ),
        default_field="created_at",
    )

    result = paginate_query(
        query,
        request=request,
        response=response,
        page=page,
        page_size=page_size,
        sort_by=sort_by,
        sort_order=sort_order,
    )
    return serialize_collection_result(result)


@router.get(
    "/{run_id}",
    response_model=AgentRunDetail,
    summary="Get one user-owned agent workflow",
)
def read_run(
    run_id: int,
    db: Session = Depends(get_db),
) -> AgentRunDetail:
    return run_detail(get_agent_run(db, run_id))


@router.post(
    "/{run_id}/retry",
    response_model=AgentRunDetail,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new attempt for a failed, cancelled, or partial workflow",
)
def retry_run(
    run_id: int,
    db: Session = Depends(get_db),
) -> AgentRunDetail:
    return run_detail(retry_agent_run(db, run_id))


@router.post(
    "/{run_id}/execute",
    response_model=AgentRunDetail,
    summary="Execute a planned multi-agent workflow",
)
def execute_run(
    run_id: int,
    data: AgentRunExecuteRequest,
    db: Session = Depends(get_db),
) -> AgentRunDetail:
    return run_detail(execute_agent_run(db, run_id, data))


@router.post(
    "/{run_id}/cancel",
    response_model=AgentRunCancelResponse,
    summary="Cancel a planned or active multi-agent workflow",
)
def cancel_run(
    run_id: int,
    db: Session = Depends(get_db),
) -> AgentRunCancelResponse:
    run = cancel_agent_run(db, run_id)
    return AgentRunCancelResponse(
        message="Agent run cancelled successfully.",
        run=run_detail(run),
    )


@router.delete(
    "/{run_id}",
    response_model=AgentRunDeleteResponse,
    summary="Delete a non-running agent workflow",
)
def remove_run(
    run_id: int,
    db: Session = Depends(get_db),
) -> AgentRunDeleteResponse:
    delete_agent_run(db, run_id)
    return AgentRunDeleteResponse(
        message="Agent run deleted successfully.",
        deleted_run_id=run_id,
    )

File: frontend/lib/agent-runs.ts
import {
  apiFetch,
  buildCollectionUrl,
  readListResponse,
  requireApiSuccess,
  type ListResult,
} from "@/lib/api";
import type {
  AgentDefinition,
  AgentRunCancelResponse,
  AgentRunCreateInput,
  AgentRunDeleteResponse,
  AgentRunDetail,
  AgentRunExecuteInput,
  AgentRunStatus,
  AgentRunSummary,
  ExecutionMode,
} from "@/types/agent-runs";

export type AgentRunListQuery = {
  page?: number;
  pageSize?: number;
  search?: string;
  status?: AgentRunStatus | "";
  executionMode?: ExecutionMode | "";
  sortBy?: "id" | "status" | "execution_mode" | "created_at" | "updated_at";
  sortOrder?: "asc" | "desc";
};

export async function listAgents(): Promise<AgentDefinition[]> {
  const response = await apiFetch("/api/agents/", {
    cache: "no-store",
  });
  await requireApiSuccess(response, "The agent registry could not be loaded.");
  return (await response.json()) as AgentDefinition[];
}

export async function listAgentRuns(
  query: AgentRunListQuery = {}
): Promise<ListResult<AgentRunSummary>> {
  const url = buildCollectionUrl("/api/agent-runs/", {
    page: query.page,
    pageSize: query.pageSize,
    search: query.search,
    status: query.status,
    execution_mode: query.executionMode,
    sortBy: query.sortBy ?? "created_at",
    sortOrder: query.sortOrder ?? "desc",
  });

  const response = await apiFetch(url, {
    cache: "no-store",
  });
  return readListResponse<AgentRunSummary>(response);
}

export async function getAgentRun(runId: number): Promise<AgentRunDetail> {
  const response = await apiFetch(`/api/agent-runs/${runId}`, {
    cache: "no-store",
  });
  await requireApiSuccess(response, "The agent workflow could not be loaded.");
  return (await response.json()) as AgentRunDetail;
}

export async function createAgentRun(
  input: AgentRunCreateInput
): Promise<AgentRunDetail> {
  const response = await apiFetch("/api/agent-runs/", {
    method: "POST",
    body: JSON.stringify(input),
  });
  await requireApiSuccess(response, "The workflow plan could not be created.");
  return (await response.json()) as AgentRunDetail;
}

export async function executeAgentRun(
  runId: number,
  input: AgentRunExecuteInput
): Promise<AgentRunDetail> {
  const response = await apiFetch(`/api/agent-runs/${runId}/execute`, {
    method: "POST",
    body: JSON.stringify(input),
  });
  await requireApiSuccess(response, "The workflow could not be executed.");
  return (await response.json()) as AgentRunDetail;
}

export async function cancelAgentRun(
  runId: number
): Promise<AgentRunCancelResponse> {
  const response = await apiFetch(`/api/agent-runs/${runId}/cancel`, {
    method: "POST",
  });
  await requireApiSuccess(response, "The workflow could not be cancelled.");
  return (await response.json()) as AgentRunCancelResponse;
}

export async function retryAgentRun(runId: number): Promise<AgentRunDetail> {
  const response = await apiFetch(`/api/agent-runs/${runId}/retry`, {
    method: "POST",
  });
  await requireApiSuccess(response, "A retry workflow could not be created.");
  return (await response.json()) as AgentRunDetail;
}

export async function deleteAgentRun(
  runId: number
): Promise<AgentRunDeleteResponse> {
  const response = await apiFetch(`/api/agent-runs/${runId}`, {
    method: "DELETE",
  });
  await requireApiSuccess(response, "The workflow could not be deleted.");
  return (await response.json()) as AgentRunDeleteResponse;
}
